### DISCUSSION TITLE
Rerun onboarding OAuth2 flow

### DISCUSSION THREAD

**Discussion by sander1988** [home-assistant/architecture#1126] [General] [answered]
I maintain several (custom) HA components and we have run into an issue regarding OAuth2. The server sometimes invalidates the token and the only way the get it renewed is by providing the credentials again. I have debugged this and this caused on the serverside, not a HA issue.

This requires the user to remove and re-add the integration so they can rerun the onboarding OAuth2 flow. Can we somehow implement a method/callback to rerun the OAuth2 flow without requiring them to re-add the integration? This will improve the user experience.

**Comment by mashdk** [comment]
A user with the issue here.
This would be a welcome feature in HA, as removing and re-adding the integration results in new Entity IDs and therefore automations, that no longer work.

**Comment by joostlek** [comment]
Please note that this is a place for architectural proposals and discussions, where the proposals are of a sufficient technological level and the intention of the author is to implement it.

For feature requests, please refer to the forums.

**Comment by sander1988** [comment]
I'm happy to implement this in HA and propose the change. So it's not really a feature request. First I would like to know if there are any objections in such a change.

Is there any technical reason why it's currently not available or reasons why we shouldn't allow reruns of the onboarding OAuth flow?

**Reply by joostlek** [reply]
Currently this proposal is a feature request, it does not show any insights in how you're going to solve this. Currently it's already possible to remove the current application_credentials, allowing you to set up the integration again using new client_id and secret.

So please add more information about how this feature would work and how it could be used

**Reply by allenporter** [reply, marked as answer]
Yes, my impression is that the existing reauth flows can support this, and it's possible to extend the auth implementation quite a bit.  Maybe moving to a discord developer support thread would be the way to go, and/or reading the dev docs on reauth.
